File: Problem 16 - Power Digit Sum.cpp
/*
Project: Project Euler - Problem 16: Power Digit Sum
Author: Mandeep Bhutani
Date: 09/19/2016

Problem:
2^15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.
What is the sum of the digits of the number 2^1000?

Description: The shortest way to find the sum of digits in an integer
is to convert the integer into a string and then
add the digits of the string into a list.
Once in a list the digits need to be converted back
to an integer and then summed. With list
comprehension this can all be achieved in a single line.
*/
#include <iostream>
#include <math.h>
#include <string>

int main() {
    int sum = 0;
    std::string number = std::to_string(pow(2, 1000));
    for (int i = 0; i < number.length(); ++i) {
        int temp = number[i] - '0'; // Subtract '0' is a way of converting from char to int
        sum += temp;
    }
    std::cout << sum + 2 << std::endl; // Add 2 in order to account for pow's double precision
}

File: problem_16_power_digit_sum.rs
/*
Project: Project Euler - Problem 16: Power Digit Sum
Author: Mandeep Bhutani
Date: 10/25/2016

Problem: 2^15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.
What is the sum of the digits of the number 2^1000?
*/
extern crate num;

use num::BigUint;
use num::Integer;
use num::ToPrimitive;
use num::pow;


fn main() {
    let mut number: BigUint = pow(BigUint::from(2 as u8), 1000);
    let mut digits: Vec<u32> = Vec::new();
    while number > BigUint::from(0 as u8) {
        let division = Integer::div_rem(&number, &BigUint::from(10 as u32));
        digits.push(ToPrimitive::to_u32(&division.1).unwrap());
        number = division.0;
    }
    let summation: u32 = digits.iter().sum();
    println!("{}", summation);
}